TrafficSimulator/Assets/Car/AutoDrive.cs

```diff
@@ -165,33 +177,94 @@ void Q_UpdateTarget()
                     _vehicleController.maxSpeedForward = _originalMaxSpeed;
                     _vehicleController.throttleInput = 1f;
+                    _vehicleController.brakeInput = 0f;
                 }
             }
             // If the vehicle is driving and the target is in front of us and we are close enough
             else if (_status == Status.Driving && dot > 0 && direction.magnitude <= _targetLookaheadDistance)
             {
                 // Set the target to the next point in the lane
-                _targetIndex = (_targetIndex + 1) % _lane.Count;
+                int nextTargetIndex = (_targetIndex + 1 + _lane.Count) % _lane.Count;
+                if (_targetIndex >= _lane.Count -1)
+                {
+                    _targetIndex = _roadEndBehaviour == RoadEndBehvaiour.Stop ? _targetIndex : nextTargetIndex;
+                }
+                else
+                {
+                    _targetIndex = nextTargetIndex;
+                }
                 _target = _lane[_targetIndex];
             }
+
+            // If the vehicle is closer to the target than the brake distance, brake
+            print(Vector3.Distance(transform.position, Q_GetBrakeTarget()));
+            if (Vector3.Distance(transform.position, Q_GetBrakeTarget()) <= _brakeDistance)
+            {
+                _vehicleController.brakeInput = 1f;
+                _vehicleController.throttleInput = 0f;
+            }
+            // If the vehicle is further away from the target than the brake distance, accelerate
+            else if (Vector3.Distance(transform.position, Q_GetBrakeTarget()) > _brakeDistance)
+            {
+                _vehicleController.brakeInput = 0f;
+                _vehicleController.throttleInput = 1f;
+            }
+        }
+
+        void Q_UpdateBrakeTarget()
+        {
+            // If the brake target is a stop target, do nothing
+            /*
+            if (_brakeTarget.type = TargetType.Stop)
+            {
+                return;
+            }
+            */
+            // Set the brake target point to the point closest to the target that is at least _brakeDistance points away
+            while (Vector3.Distance(transform.position, Q_GetBrakeTarget()) < _brakeDistance)
+            {
+                if (_brakeTargetIndex + 1 >= _lane.Count)
+                {
+                    return;
+                }
+                _brakeTargetIndex = (_brakeTargetIndex + 1 + _lane.Count) % _lane.Count;
+                _brakeTarget = _lane[_brakeTargetIndex];
+            }
+            
+        }
+
+        void Q_UpdateBrakeDistance()
+        {
+            // If the vehicle is moving
+            if (_vehicleController.speed > 0)
+            {
+                // Calculate the distance it will take to stop
+                _brakeDistance = 5 + (_vehicleController.speed / 2) + _vehicleController.speed * _vehicleController.speed / (_vehicleController.tireFriction * 9.81f);
```

Comment:
Random addition with 5, should create constant for it in order to more easily understand why it is there